### PR TITLE
feat(bts): Implement logging for event data as Json string

### PR DESCRIPTION
**Pull request by shreyasbhat0**

### CODE REVIEW

File: rust/near/bsh/bts/src/coin_management.rs [hunk 1]
```diff
@@ -119,12 +140,29 @@ impl BtpTokenService {
...
                     .set(&env::current_account_id(), &coin_id, balance);
-
-                log!("[Burn] {} {}", amount, coin_symbol);
+                let coin_name = self.coins.get(&coin_id).unwrap().name().to_string();
+                log!(json!(
+                {
+                    "event":"Burn",
+                    "code":1,
+                    "amount":amount,
+                    "token_name":coin_name
```
Comment (MuhammedIrfan) [inline rust/near/bsh/bts/src/coin_management.rs:149]:
```suggestion
                    "amount": amount,
                    "token_name": coin_name
```

File: rust/near/bsh/bts/src/coin_management.rs [hunk 2]
```diff
@@ -119,12 +140,29 @@ impl BtpTokenService {
...
             PromiseResult::NotReady => log!("Not Ready"),
             PromiseResult::Failed => {
-                log!("[Burn Failed] {} {}", amount, coin_symbol);
+                let coin_name = self.coins.get(&coin_id).unwrap().name().to_string();
+                log!(json!(
+                {
+                    "event":"Burn",
+                    "code":0,
+                    "amount":amount,
+                    "token_name":coin_name
```
Comment (MuhammedIrfan) [inline rust/near/bsh/bts/src/coin_management.rs:162]:
```suggestion
                    "amount": amount,
                    "token_name": coin_name
```

File: rust/near/bsh/bts/src/messaging.rs [hunk 3]
```diff
@@ -75,22 +75,46 @@ impl BtpTokenService {
...
-                    receiver,
-                    serial_no,
-                    assets
-                ),
+                PromiseResult::Successful(_) => {
+                    let mut assets_log: Vec<Value> = Vec::new();
+                    assets.iter().for_each(|asset| {
+                        assets_log.push(json!({
+                        "token_name": asset.name(),
+                        "amount":asset.amount(),
```
Comment (MuhammedIrfan) [inline rust/near/bsh/bts/src/messaging.rs:83]:
```suggestion
                        "amount": asset.amount(),
```

File: rust/near/bsh/bts/src/messaging.rs [hunk 4]
```diff
@@ -75,22 +75,46 @@ impl BtpTokenService {
...
+                        "fee": asset.fees(),
+                        }))
+                    });
+                    let log = json!({
+                      "event":"TransferStart",
```
Comment (MuhammedIrfan) [inline rust/near/bsh/bts/src/messaging.rs:88]:
```suggestion
                      "event": "TransferStart",
```

File: rust/near/bsh/bts/src/messaging.rs [hunk 5]
```diff
@@ -75,22 +75,46 @@ impl BtpTokenService {
...
+                      "code":1,
+                      "sender_address":sender,
+                      "serial_number":serial_no,
+                      "receiver_address":receiver,
+                      "assets":assets_log
```
Comment (MuhammedIrfan) [inline rust/near/bsh/bts/src/messaging.rs:93]:
```suggestion
                      "sender_address": sender,
                      "serial_number": serial_no,
                      "receiver_address": receiver,
                      "assets": assets_log
```

File: rust/near/bsh/bts/src/messaging.rs [hunk 6]
```diff
@@ -75,22 +75,46 @@ impl BtpTokenService {
...
-                        sender,
-                        receiver,
-                        serial_no,
-                        assets
-                    );
+                    let mut assets_log: Vec<Value> = Vec::new();
+                    assets.iter().for_each(|asset| {
+                        assets_log.push(json!({
+                        "token_name": asset.name(),
+                        "amount":asset.amount(),
```
Comment (MuhammedIrfan) [inline rust/near/bsh/bts/src/messaging.rs:104]:
```suggestion
                        "amount": asset.amount(),
```

File: rust/near/bsh/bts/src/accounting.rs [hunk 7]
```diff
@@ -160,22 +160,40 @@ impl BtpTokenService {
...
-                    account,
-                    coin_symbol
-                );
+                log!(json!(
+                {
+                    "event":"Withdraw",
+                    "status":"Success",
+                    "by":account,
+                    "amount":amount,
+                    "token_name":coin_symbol
```
Comment (MuhammedIrfan) [inline rust/near/bsh/bts/src/accounting.rs:169]:
```suggestion
                    "by": account,
                    "amount": amount,
                    "token_name": coin_symbol
```

File: rust/near/bsh/bts/src/coin_management.rs [hunk 8]
```diff
@@ -98,12 +98,33 @@ impl BtpTokenService {
                     .set(&env::current_account_id(), &coin_id, balance);
 
                 self.internal_transfer(&env::current_account_id(), &receiver_id, &coin_id, amount);
+                let coin_name = self.coins.get(&coin_id).unwrap().name().to_string();
 
-                log!("[Mint] {} {}", amount, coin_symbol);
+                log!(json!(
+                {
+                    "event":"Mint",
```
Comment (MuhammedIrfan) [inline rust/near/bsh/bts/src/coin_management.rs:105]:
```suggestion
                    "event": "Mint",
```

File: rust/near/bsh/bts/src/coin_management.rs [hunk 9]
```diff
@@ -119,12 +140,29 @@ impl BtpTokenService {
                 balance.deposit_mut().sub(amount).unwrap();
                 self.balances
                     .set(&env::current_account_id(), &coin_id, balance);
-
-                log!("[Burn] {} {}", amount, coin_symbol);
+                let coin_name = self.coins.get(&coin_id).unwrap().name().to_string();
+                log!(json!(
+                {
+                    "event":"Burn",
```
Comment (MuhammedIrfan) [inline rust/near/bsh/bts/src/coin_management.rs:146]:
```suggestion
                    "event": "Burn",
```

File: rust/near/bsh/bts/src/coin_management.rs [hunk 10]
```diff
@@ -119,12 +140,29 @@ impl BtpTokenService {
...
+                .as_str()
+                .unwrap());
             }
             PromiseResult::NotReady => log!("Not Ready"),
             PromiseResult::Failed => {
-                log!("[Burn Failed] {} {}", amount, coin_symbol);
+                let coin_name = self.coins.get(&coin_id).unwrap().name().to_string();
+                log!(json!(
+                {
+                    "event":"Burn",
```
Comment (MuhammedIrfan) [inline rust/near/bsh/bts/src/coin_management.rs:159]:
```suggestion
                    "event": "Burn",
```

**Review by MuhammedIrfan — CHANGES_REQUESTED** [review]

**Comment by sanglt1902** [comment]
@MuhammedIrfan LGTM, can you help to deploy it ASAP and perform to transfer of some transactions?

**Review by shreyasbhat0 — COMMENTED** [review]
Changes has been made

File: rust/near/bsh/bts/src/accounting.rs [hunk 11]
```diff
@@ -160,22 +160,33 @@ impl BtpTokenService {
...
-                log!(
-                    "[Withdraw Failed] Amount : {} by {}  {}",
-                    amount,
-                    account,
-                    coin_symbol
-                );
+                log!(json!(
+                {
+                    "event": "Withdraw",
+                    "code": 0,
```
Comment (MuhammedIrfan) [inline rust/near/bsh/bts/src/accounting.rs:183]:
```suggestion
                    "code": 1,
```

File: rust/near/bsh/bts/src/messaging.rs [hunk 12]
```diff
@@ -75,22 +75,46 @@ impl BtpTokenService {
...
+                    let log = json!({
+                      "event": "TransferStart",
+                      "code": 1,
+                      "sender_address": sender,
+                      "serial_number": serial_no,
+                      "receiver_address": receiver,
+                      "assets" : assets_log
+                    });
+
+                    log!(log.as_str().unwrap().to_string());
```
Comment (MuhammedIrfan) [inline rust/near/bsh/bts/src/messaging.rs:117]:
```suggestion
                    log!(log.as_str().unwrap());
```

File: rust/near/bsh/bts/src/messaging.rs [hunk 13]
```diff
@@ -75,22 +75,46 @@ impl BtpTokenService {
...
+                    let log = json!({
+                      "event": "TransferStart",
+                      "code": 0,
+                      "sender_address": sender,
+                      "serial_number": serial_no,
+                      "receiver_address": receiver,
+                      "assets": assets_log
+                    });
+
+                    log!(log.as_str().unwrap().to_string())
```
Comment (MuhammedIrfan) [inline rust/near/bsh/bts/src/messaging.rs:96]:
```suggestion
                    log!(log.as_str().unwrap())
```

**Review by MuhammedIrfan — APPROVED** [review]

**Comment by MuhammedIrfan** [comment]
> @MuhammedIrfan LGTM, can you help to deploy it ASAP and perform to transfer of some transactions?

Sure

**Review by MuhammedIrfan — CHANGES_REQUESTED** [review]
Please fix the unit tests

**Comment by shreyasbhat0** [comment]
> Please fix the unit tests

this has been fixed.

**Review by MuhammedIrfan — APPROVED** [review]

**Pull request merged**